src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2055,458 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user and the shard metadata is
+		 * not sane, the user can only affect its own tables. Given that the
+		 * user is owner of the table, we should allow.
+		 */
+		EnsureShardMetadataIsSane(relationId, shardId, storageType, shardMinValue,
+								  shardMaxValue);
+	}
+
+	InsertShardRow(relationId, shardId, storageType, shardMinValue, shardMaxValue);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureCoordinatorInitiatedOperation is a helper function which ensures that
+ * the execution is initiated by the coordinator on a worker node.
+ */
+static void
+EnsureCoordinatorInitiatedOperation(void)
+{
+	/*
+	 * We are restricting the operation to only MX workers with the local group id
+	 * check. The other two checks are to ensure that the operation is initiated
+	 * by the coordinator.
+	 */
+	if (!IsCitusInitiatedRemoteBackend() || !MyBackendIsInDisributedTransaction() ||
+		GetLocalGroupId() == COORDINATOR_GROUP_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("This is an internal function that only Citus "
+							   "requires to use in a distributed transaction")));
+	}
+}
+
+
+/*
+ * EnsureShardMetadataIsSane ensures that the input values are safe
+ * for inserting into pg_dist_shard metadata.
+ */
+static void
+EnsureShardMetadataIsSane(Oid relationId, int64 shardId, char storageType,
+						  text *shardMinValue, text *shardMaxValue)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+
+	if (!(storageType == SHARD_STORAGE_TABLE ||
+		  storageType == SHARD_STORAGE_FOREIGN ||
+		  storageType == SHARD_STORAGE_COLUMNAR))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard storage type: %c", storageType)));
+	}
+
+	char partitionMethod = PartitionMethodViaCatalog(relationId);
+	if (partitionMethod == DISTRIBUTE_BY_INVALID)
+	{
+		/* connection from the coordinator operating on a shard */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("The relation \"%s\" does not have a valid "
+							   "entry in pg_dist_partition.",
+							   get_rel_name(relationId))));
+	}
+	else if (!(partitionMethod == DISTRIBUTE_BY_HASH ||
+			   partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		/* connection from the coordinator operating on a shard */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Metadata syncing is only allowed for hash, "
+							   "reference and local tables: %c", partitionMethod)));
+	}
+
+	List *distShardTupleList = LookupDistShardTuples(relationId);
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		if (!(shardMinValue == NULL && shardMaxValue == NULL))
```

Comment:
nit: I have slight prefererence for this way of writing this check (feel free to ignore if you disagree)
```suggestion
		if (shardMinValue != NULL || shardMaxValue != NULL))
```